convert-hf-to-gguf.py

```diff
@@ -1624,54 +1624,13 @@ class Qwen2MoeModel(Model):
     model_arch = gguf.MODEL_ARCH.QWEN2MOE
 
     def set_gguf_parameters(self):
-        self.gguf_writer.add_name(self.dir_model.name)
-        self.gguf_writer.add_block_count(self.block_count)
-
-        if (n_ctx := self.find_hparam(["max_position_embeddings", "n_ctx"], optional=True)) is not None:
-            self.gguf_writer.add_context_length(n_ctx)
-            logger.info(f"gguf: context length = {n_ctx}")
-
-        n_embd = self.find_hparam(["hidden_size", "n_embd"])
-        self.gguf_writer.add_embedding_length(n_embd)
-        logger.info(f"gguf: embedding length = {n_embd}")
-
-        n_head = self.find_hparam(["num_attention_heads", "n_head"])
-        self.gguf_writer.add_head_count(n_head)
-        logger.info(f"gguf: head count = {n_head}")
-
-        if (n_head_kv := self.hparams.get("num_key_value_heads")) is not None:
-            self.gguf_writer.add_head_count_kv(n_head_kv)
-            logger.info(f"gguf: key-value head count = {n_head_kv}")
-
-        if (rope_theta := self.hparams.get("rope_theta")) is not None:
-            self.gguf_writer.add_rope_freq_base(rope_theta)
-            logger.info(f"gguf: rope theta = {rope_theta}")
-        if (f_rms_eps := self.hparams.get("rms_norm_eps")) is not None:
-            self.gguf_writer.add_layer_norm_rms_eps(f_rms_eps)
-            logger.info(f"gguf: rms norm epsilon = {f_rms_eps}")
-        if (f_norm_eps := self.find_hparam(["layer_norm_eps", "layer_norm_epsilon", "norm_epsilon"], optional=True)) is not None:
-            self.gguf_writer.add_layer_norm_eps(f_norm_eps)
-            logger.info(f"gguf: layer norm epsilon = {f_norm_eps}")
-
-        if (n_experts_used := self.hparams.get("num_experts_per_tok")) is not None:
-            self.gguf_writer.add_expert_used_count(n_experts_used)
-            logger.info(f"gguf: experts used count = {n_experts_used}")
-
-        if (n_experts := self.find_hparam(["num_experts", "num_local_experts"])) is not None:
+        super().set_gguf_parameters()
+        if (n_experts := self.hparams.get("num_experts")) is not None:
             self.gguf_writer.add_expert_count(n_experts)
-
         if (moe_intermediate_size := self.hparams.get("moe_intermediate_size")) is not None:
             self.gguf_writer.add_expert_feed_forward_length(moe_intermediate_size)
             logger.info(f"gguf: expert feed forward length = {moe_intermediate_size}")
 
-        if (shared_expert_intermediate_size := self.find_hparam(["shared_expert_intermediate_size","intermediate_size", "n_inner"])) is not None:
-            self.gguf_writer.add_feed_forward_length(shared_expert_intermediate_size)
-            logger.info(f"gguf: feed forward length = {shared_expert_intermediate_size}")
```

Comment:
Why was this removed?

Reply:
not removed, but cleaned up. 
`add_feed_forward_length` is called by `super().set_gguf_parameters()` as it was before my PR. 

`add_expert_feed_forward_length` is the only new thing that is added, and is the one taken into account here https://github.com/ggerganov/llama.cpp/pull/7835/files#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR5814 and here https://github.com/ggerganov/llama.cpp/pull/7835/files#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR5820

Reply:
Not sure as what though, @ggerganov any thoughts?

Trying to see if there are any good docs on these values...

Reply:
I'm unsure what `intermediate_size` is either. For `Qwen1.5-MoE-A2.7B` the values look correct, that's why I'm inclined to say that `intermediate_size = 18944` could have been a mistake. 

As for storing `shared_expert_intermediate_size`, I see no specific slot for it, apart from overriding `LLM_KV_FEED_FORWARD_LENGTH` which might not be the best option (even tho' it's unused)

Reply:
`Qwen2MoE` docs on HF, not very helpful:
* **intermediate_size** (int, optional, defaults to 5632) — Dimension of the MLP representations.
* **moe_intermediate_size** (int, optional, defaults to 1408) — Intermediate size of the routed expert.
* **shared_expert_intermediate_size** (int, optional, defaults to 5632) — Intermediate size of the shared expert.

Reply:
> As for storing `shared_expert_intermediate_size`, I see no specific slot for it, apart from overriding `LLM_KV_FEED_FORWARD_LENGTH` which might not be the best option (even tho' it's unused)

You would have to make a new one.

Reply:
You'd probably want both values captured for the future, in case a model comes along where **ffn_*_exps** and **ffn_*_shexp** shapes are independently driven, but in this case (and I think for a lot of other cases too) they are in sync

Reply:
Aha, take a look here:
https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_moe/modeling_qwen2_moe.py#L860-L865

`57B-A14B` is a sparse model and `intermediate_size` is then overridden by `shared_expert_intermediate_size`:
https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_moe/modeling_qwen2_moe.py#L804

Reply:
> > As for storing `shared_expert_intermediate_size`, I see no specific slot for it, apart from overriding `LLM_KV_FEED_FORWARD_LENGTH` which might not be the best option (even tho' it's unused)
> 
> You would have to make a new one.

I'm unsure if this change should be made within the scope of this PR. I could open another PR referencing this one's discussions.

Reply:
> 57B-A14B is a sparse model and intermediate_size is then overridden by shared_expert_intermediate_size

I see, so technically, setting `LLM_KV_FEED_FORWARD_LENGTH` for `57B-A14B` to `shared_expert_intermediate_size` was a correct solution all along

Reply:
Yeah, none of them have `mlp_only_layers` yet though.

Reply:
@CISC What do you think the next steps should be?

We can leave this PR as is since it works, and then start a new one with a proper impl?

Otherwise, adding `mlp_only_layers (int[])` and `decoder_sparse_step: int` to the config (and converter) is probably out of scope.

Reply:
I think adding a `shared_expert_feed_forward_length` is within the scope of this PR.

The sparse layer stuff should be another PR.

Reply:
Ok, would you like to simply override `LLM_KV_FEED_FORWARD_LENGTH` with `shared_expert_intermediate_size` or create a new one, say `LLM_KV_SHARED_EXPERT_FEED_FORWARD_LENGTH`?

Reply:
The latter since you need to keep the original value for the other PR.

Reply:
> `mlp_only_layers: int[]`, and `decoder_sparse_step: int`

@legraphista I've had to deal with models which have both MoE layers and MLP-only layers in #7531 (Jamba). A new metadata key-value pair is not needed. The easy way is to [check for the presence of `layers.ffn_gate_inp`](https://github.com/ggerganov/llama.cpp/blob/33425a7e1ed366082a2dbf64f2485531471515e0/llama.cpp#L7756), which is only there on MoE layers, and when building the compute graph, [`build_llama`](https://github.com/ggerganov/llama.cpp/blob/963552903f51043ee947a8deeaaa7ec00bc3f1a4/llama.cpp#L7502) can be a good inspiration for how to check this per-layer.